### ISSUE TITLE
Can train and decode work with punctuation?

### ISSUE THREAD

**Issue by maxbaluev** [facebookresearch/wav2letter#466]

**Comment by tlikhomanenko** [comment]
Hi @maxbaluev,

Yes, you need to have them among tokens set. Also it is better to train language model with them too and have them in the lexicon.

**Comment by MHumza3656** [comment]
By train, do you mean train from scratch or finetune your current provided language models?

Is there a Transformer model for speech-to-text that is trained with punctuations? and only by adding ".,?!" in tokens it will start working?

**Comment by tlikhomanenko** [comment]
We didn't train with punctuation. You can use our models, change the last layer to the larger layer which maps to tokens + punctuation and then finetune on data with punctuation.

About LM: if ngram - retrain, if common crawl - you can train new one on your data and then do interpolation.

My main understanding that it is better do postprocessing with punctuation rather than ASR itself, as mostly punctuation is an LM part.

**Comment by MHumza3656** [comment]
I tried a punctuation model on top of the results provided by the w2l model. The issue with that is that punctuations are placed poorly since it's more of conversational audios, what I'm transcribing. You can understand that in these types of audio punctuation is not mainly catered very well, 2 people speak with each other or in reply to another and that breaks the normal punctuation rules I think.

Why I think ASR might work well, please correct me if I'm wrong, it has the pauses within the audio, the rise of a voice all of these could be potential features for the ",!?." that basically makes up the sentence.

**Comment by MHumza3656** [comment]
Could you help me understand a little, how do I take the last layer of the model and train it? I mean I can do that with a TensorFlow trained model but is this doable in Python? or C++?

**Comment by MHumza3656** [comment]
> So when you have loaded the model (I guess sequential) you need to create another one where Sequential of all layers except one is stored in new var net1 and last layer is stored in another var lastLayer. For net1 you set it to eval model (as we do for evaluation) and for lastLayer you set it to train mode so that grads will be computed.

I have followed the following link for this, with very slight variations since this example is outdated.
https://github.com/flashlight/wav2letter/issues/507#issuecomment-580809478
You know this code better, do you think its doing the same thing as you have mentioned above?

**Comment by tlikhomanenko** [comment]
Yep, this is good example, so you load model into tmp model and create the same model from the same arch and then set necessary params in the loop from the one you loaded.

You need to fine-tune at least last layer, because now you change the linear layer to output to another set of tokens, and it doesn't learn anything about punctuation before. 

Other solution could be if you train LM with the punctuation and you can hack beam-search decoder to infer additional punctuation using only LM predictions, but this will take a lot more efforts to do.

**Comment by MHumza3656** [comment]
Yeah seems like it would take much more time. However, I tried the first way and I believe it's working great.
It seems the ASR Model within the Finetuning script is auto-assigning the Training epochs based on the training examples. Is there a way to override that and provide it a constant number? I don't see any Hyper-parameter that addresses "Epochs".

**Comment by tlikhomanenko** [comment]
Do you want to set number of epochs to train? Then it is https://github.com/flashlight/flashlight/blob/master/flashlight/app/asr/common/Flags.cpp#L129-L132

**Comment by MHumza3656** [comment]
Any suggestions for epochs? I'm Finetuning with 25 Thousand audio examples. The model I am using is this one `am_transformer_ctc_stride3_letters_300Mparams.bin`.
GPU and time are not an issue, even if it takes a week to Train that's fine.

**Comment by tlikhomanenko** [comment]
I would set just very huge number and see how it trains. If on valid WER is on plateau then reduce lr.

**Comment by MHumza3656** [comment]
Sounds good, thank you very much :) 👍

**Comment by MHumza3656** [comment]
> Yep, this is good example, so you load model into tmp model and create the same model from the same arch and then set necessary params in the loop from the one you loaded.

@tlikhomanenko 1 more question, With this strategy, we're basically scratching the final layer and training it from scratch based on the previous layers' frozen weights right?
Is this enough for the model to learn? or are the previous weights not frozen? They're being trained further (Fine-tuned) as well?
If the above is not the case, should more layers be added and then the final layer be added in front of them? i.e. I have 10 layers initially, I scratched the 10th layer (Output) and finetuned - This is the first scenario. **2nd Scenario** I have 10 layers, I deleted (Output) added 2-3 more layers, then attached Output. Now, I have 12-13 layers. Is this possible and do you think this will provide more effective Finetuning?

**Comment by MHumza3656** [comment]
alright makes sense. but 
> otherwise try to finetune the whole not.

That would be quite a costly/time-consuming thing, right? Shall I unfreeze a few of the layers? So, if there are a total of 20 layers I maybe unfreeze 2-3 layers. Is that a good thing to do? If yes, then which layers do you suggest?

**Comment by tlikhomanenko** [comment]
I would say it should not be costly as potentially only last layer should be adapted a lot but other layers only a bit. So I would suggest do finetuning only with last layer while the rest are frozen and then do lr decay with finetuning all layers, or doing a bit of training with finetuning all layers, so that the whole net and last layer are making necessary final adjustments.
